### DISCUSSION TITLE
Is there an existing solution for egui + SDL2 + wgpu combo?

### DISCUSSION THREAD

**Discussion by kaphula** [emilk/egui#1690] [Q&A] [answered]
I am using SDL2 and wgpu in my project so can I use these dependencies automatically to handle the events and rendering of egui?

**Comment by zicklag** [comment]
You could look at the community [SDL2 backend](https://github.com/ArjunNair/egui_sdl2_gl) for egui, but I don't know that you can combine it with wgpu. I don't know much about sdl2.

**Reply by kaphula** [reply]
I used this earlier successfully but then I switched to wgpu from pure OpenGL. I tried to use it the way as before alongside with wgpu but I could not get my project to compile because bytemuck went somehow crazy when I enabled `egui` and `egui_sdl2_gl` crates in my new project configuration. 

I think I now managed to fix this project error by adding proper derive feature to bytemuck, so I'll have to try if I can get the old setup working (`gl` + `egui` + `egui_sdl2_gl`) alongside with wgpu. I'll post the results here.

**Comment by kaphula** [comment]
Unfortunately `egui_sdl2_gl` fails to compile the shaders of the painter when this line is called in my app (I debugged inside what causes the crash and it is the vertex shader compilation): 
```rust 
    let (mut painter, mut egui_state) = egui_sdl2_gl::with_sdl2(&engine.sdl_system.window, ShaderVersion::Default, DpiScaling::Default);
```

OpenGL does not give any error string from the compilation, just empty string.

I am using these version now:
```toml
egui = "0.16.0"
egui_sdl2_gl = "0.16.0"
```

I think I used `0.15` earlier but now I have to use `0.16` because older versions of `egui_sdl2_gl` forces a use of older version of `sdl2-rust` where the `raw-window-handle` required by `wgpu` is not yet supported.

**Comment by kaphula** [comment, marked as answer]
Well okay I got it to render the GUI elements on top of my app but it is not good. The whole view flickers and if I try to resize the window the application freezes or goes crazy.

**Reply by zicklag** [reply]
That's no good. I wonder if it's just the `egui_sdl2_gl` crate not working well. I feel like it should be possible to use the `egui_glow` backend to do the rendering on top of the raw OpenGL context, even if you use WGPU, but I've never used SDL2, so I'm not exactly sure what that'd look like.

**Reply by kaphula** [reply]
Perhaps but I guess the problem lies in SDL2's `window.gl_swap_window()` and wgpu's final render submit. Can you guarantee that these render calls will be called always in the same order when 2 different APIs are being used? Or perhaps it is something else related to clearing OpenGL color and depth buffers for example.

I looked into the existing options of egui and to me it seems like one solution that could be reasonable to implement is by taking the window input code from `egui_sdl2_gl` and the egui rendering code from `egui_wgpu_backend` and then combine these to make a crate that would allow the use of egui when SDL2 is being used as the windowing backend for a wgpu application. I may attempt to stitch these together at some point on my own. If you think there's some obvious issue with this approach let me know.

**Reply by zicklag** [reply]
That idea sounds like it could work. If you just do all your rendering with WGPU, and you just pass the raw window handle from SDL2 to WGPU, it seems like that could be fine.

**Reply by kaphula** [reply]
I got this pretty much working now, at least to the extend that satisfies my needs. I am planning to separate the SDL2+wgpu implementation to its own crate and a test application that uses it to its own crate. These crates are now living in the workspace of egui in my fork. The combined implementation is [here](https://github.com/kaphula/egui/blob/sdl2-wgpu/examples/hello_sdl2_wgpu/src/main.rs).

This could be merged to egui later if they are interested in having this kind of edge case example and implementation included in the main repository. What do you think?

**Reply by zicklag** [reply]
I'm not sure, that depends on @emilk 's preference. I imagine if it was useful to you, it could be useful to other people, too. But I think maybe a separate repo makes sense. I'm not sure if SDL2 is something we want to maintain in the main Egui repo or not.

**Reply by kaphula** [reply]
Sounds good. I created a [separate repository](https://github.com/kaphula/egui-sdl2-event) and crate out of this case.
